### DISCUSSION TITLE
standData是福还是祸？

### DISCUSSION THREAD

**Discussion by wjgogogo** [antvis/S2#384] [General]
从注释来看：引入标准数据为了转换提高效率，然而🤔 ：
![image](https://user-images.githubusercontent.com/17964556/135116211-88d5e191-5245-47c8-8370-f9cb3a026b16.png)
![image](https://user-images.githubusercontent.com/17964556/135116350-a5ff9e34-5c87-46f0-88de-4c8f82646456.png)

目前标准数据流导致的问题：
* 以前是按照`values`遍历，保证了生成的数据都是以`values`为准，而现在的指标顺序和**数据的内部顺序**强关联，甚至出现如果没有一个对应字段的数据，根本不会在最终结果显示该 value 的单元格，和`values`配置不匹配。
*  data数据处理和下钻处理数据方法不匹配：
![image](https://user-images.githubusercontent.com/17964556/135117695-1aa05401-7db9-4e59-a7d2-374263d34164.png)

当然可行的解决方法也有：
1. 可以再写一个工具方法来处理原始数据不存在某些字段的情况，但是我们的初衷是提高转换速度，这样只会加大难度💀。
2. 或者在生成`rowPivot`或者`colPivot`做补全处理，又得多`if/else`几层判断
3. 最简单的方法就是用回原来的转换流程，考虑到现在没几个用户在用，成本也低

相关 issues:  #224, #292 

各位怎么看这个问题？

**Comment by xingwanying** [comment]
我站提供工具函数外部转换。
理由如下：

1、我们最终的标准就是一个格子对应一个数据，`dirllDown` 只能说是去除 `standData` 没有挪干净。
2、增加一个 `standData` 的概念增加了用户的理解成本。
3、 我们不能预判用户的数据是怎样的，因为很有可能用户的默认数据就是 `standData`, 或者是其他的各种格式，即使用回原来也没办法保证用户不需要进行一次格式转换，我们没办法涵盖所有场景，之前纯粹是按照我们自己的场景来适配的。
4、把这层逻辑放在 `core` 层反而是增加了性能负担。

**Reply by wjgogogo** [reply]
是的，关键在于我们提供的工具函数也得说明其针对的场景，毕竟这个逻辑不管是放在外部还是内部，都没法处理所有情况。
首要是将数据的问题导致 `values` 没有完全显示的问题解了， 现在不是按照`values`的配置排序的

**Reply by xingwanying** [reply]
不是按照 `values` 的配置排序的? 具体指的是什么？可以描述详细一些嘛？

**Reply by wjgogogo** [reply]
举个🌰  :
1. `values`配置是 `['price', 'cost']`
2. `data`为：
```js
[
    {
        "province": "浙江",
        "city": "杭州",
        "type": "笔",
        "cost": "1"
    },
    {
        "province": "浙江",
        "city": "杭州",
        "type": "笔",
        "price": "2"
    }
]
```
3. 根据 standard 上面截图的处理来看，会生成：
```js
[
    {
        "province": "浙江",
        "city": "杭州",
        "type": "笔",
        "cost": "1",
        "extra-field": "cost",
        "field-value": "1",
    },
    {
        "province": "浙江",
        "city": "杭州",
        "type": "笔",
        "price": "2",
        "extra-field": "price",
        "field-value": "2",

    }
]
```
4. 最终会按data顺序生成`sortedDimensionValues`时，导致 `cost`在 `price`之前 ：
https://github.com/antvis/S2/blob/cfc77b7823657897b560c182362af26c0205b47b/packages/s2-core/src/utils/dataset/pivot-data-set.ts#L119-L134

**Reply by xingwanying** [reply]
ref: https://github.com/antvis/S2/pull/528
